keras_nlp/layers/bert_packer.py

```diff
@@ -0,0 +1,230 @@
+
+    Args:
+        sequence_length: The desired output length.
+        start_value: The id or token that is to be placed at the start of each
+            sequence (called "[CLS]" for BERT). The dtype much mach the dtype of
+            the input tensors to the layer.
+        end_value: The id or token that is to be placed at the end of each
+            input segment (called "[SEP]" for BERT). The dtype much mach the
+            dtype of the input tensors to the layer.
+        pad_value: The id or token that is to be placed into the unused
+            unused positions after the last segment in the sequence
+            (called "[PAD]" for BERT).
+        truncator: The algorithm to truncate a list of batched segments to fit a
+            per-example length limit. The value can be either `round_robin` or
+            `waterfall`:
+                - `"round_robin"`: Available space is assigned one token at a
+                    time in a round-robin fashion to the inputs that still need
+                    some, until the limit is reached.
+                - `"waterfall"`: The allocation of the budget is done using a
+                    "waterfall" algorithm that allocates quota in a
+                    left-to-right manner and fills up the buckets until we run
+                    out of budget. It support arbitrary number of segments.
+
+    Examples:
+
+    *Pack a single input for classification.*
+    >>> seq1 = [1, 2, 3, 4]  # Replace with a real input.
+    >>> packer = keras_nlp.layers.BertPacker(8, start_value=101, end_value=102)
+    >>> packer(seq1)
+    {
+        "tokens": <tf.Tensor: shape=(8,), dtype=int32,
+            numpy=array([101,   1,   2,   3,   4, 102,   0,   0], dtype=int32)>,
+        "padding_mask": <tf.Tensor: shape=(8,), dtype=float32,
+            numpy=array([1., 1., 1., 1., 1., 1., 0., 0.], dtype=float32)>,
+        "segment_ids": <tf.Tensor: shape=(8,), dtype=float32,
+            numpy=array([0., 0., 0., 0., 0., 0., 0., 0.], dtype=float32)>,
+    }
+
+    *Pack multiple inputs for classification.*
+    >>> seq1 = [1, 2, 3, 4]  # Replace with a real input.
+    >>> seq2 = [11, 12, 13, 14]  # Replace with a real input.
+    >>> seq3 = [21, 12, 23, 24]  # Replace with a real input.
+    >>> packer = keras_nlp.layers.BertPacker(8, start_value=101, end_value=102)
+    >>> packer((seq1, seq2, seq3))
+    {
+        "tokens": <tf.Tensor: shape=(8,), dtype=int32,
+            numpy=array([101,   1,   2, 102,  11, 102,  21,  102], dtype=int32)>,
+        "padding_mask": <tf.Tensor: shape=(8,), dtype=float32,
+            numpy=array([1., 1., 1., 1., 1., 1., 1., 1.], dtype=float32)>,
+        "segment_ids": <tf.Tensor: shape=(8,), dtype=float32,
+            numpy=array([0., 0., 0., 0., 1., 1., 2., 2.], dtype=float32)>,
+    }
+
+    Reference:
+        [Devlin et al., 2018](https://arxiv.org/abs/1810.04805).
+    """
+
+    def __init__(
+        self,
+        sequence_length,
+        start_value,
```

Comment:
I went with value because this is could be either a string token or an int id (depending on input type), so I figured value was a good middle ground. Open to other naming choices.

Reply:
I think token can mean both as well? This is just personal thought, but "token" is more straightforward, also HF is using "token": https://huggingface.co/transformers/v2.9.1/main_classes/tokenizer.html

Reply:
The API you are linking is for string tokens, not the ids they correspond to. Hugging face call the integer side `ids`, e.g. `tokenizer.bos_token_id`.

We just need to pick conversions and be consistent. In the MLM layer we called arguments `mask_token_id` and `unselectable_token_ids`.

The dominant use case for this layer will be with ints not strings. So if anything start_token_id is better than start_token. But this is a nice ambiguous middle ground, not necessarily a token_id, not necessarily a token. I think we should name it reflecting that.